File: tests/real-media.test.ts
/**
 * 実音声映像の照合（TypeScript、段 A の実データ版）。
 *
 * 何を証明するか: 合成したベクタではなく、**実際に符号化された AV1 と Opus** に対して、
 * ワイヤ形式の符号化・復号が往復で一致し、破棄可否と破棄優先順位の判断が凍結資産と
 * 一致すること。同じ資産を 6 言語すべてが照合する（言語間の差はここで露出する）。
 *
 * 資産の作り方は tools/real-media.ts にある。資産を実装に合わせて変更してはならない（ADR-0012）。
 */

import { test } from "node:test";
import assert from "node:assert/strict";
import { readFileSync } from "node:fs";

import {
  computeDiscardable,
  dropPriority,
  decodeMediaMessage,
  encodeMediaMessage,
  type MediaMessage,
} from "../packages/core/src/wire.ts";
import { CHANNEL_AUDIO, CHANNEL_VIDEO } from "../packages/core/src/generated/wire-layout.ts";

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

function hexToBytes(hex: string): Uint8Array {
  const bytes = new Uint8Array(Math.trunc(hex.length / 2));
  for (let index = 0; index + 1 < hex.length; index += 2) {
    bytes[Math.trunc(index / 2)] = Number.parseInt(hex.slice(index, index + 2), 16);
  }
  return bytes;
}

function bytesToHex(bytes: Uint8Array): string {
  let out = "";
  for (const byte of bytes) {
    out += byte.toString(16).padStart(2, "0");
  }
  return out;
}

function readAsset(): Record<string, unknown> {
  const text = readFileSync(new URL("../spec/vectors/real-media.json", import.meta.url), "utf8");
  const value: unknown = JSON.parse(text);
  assert.ok(isRecord(value), "資産が連想配列である");
  return value;
}

function numberOf(record: Record<string, unknown>, key: string): number {
  const value = record[key];
  assert.equal(typeof value, "number", `${key} が数値である`);
  return typeof value === "number" ? value : 0;
}

function textOf(record: Record<string, unknown>, key: string): string {
  const value = record[key];
  assert.equal(typeof value, "string", `${key} が文字列である`);
  return typeof value === "string" ? value : "";
}

function boolOf(record: Record<string, unknown>, key: string): boolean {
  const value = record[key];
  assert.equal(typeof value, "boolean", `${key} が真偽値である`);
  return value === true;
}

test("実 AV1 の資産が空でなく、キーフレームと時間層 3 層を含む", () => {
  const asset = readAsset();
  const video = asset["video"];
  assert.ok(isRecord(video));
  const frames = video["frames"];
  assert.ok(Array.isArray(frames));
  assert.ok(frames.length >= 30, `映像が 30 枚以上ある（実際 ${frames.length}）`);
  assert.equal(numberOf(video, "temporalLayers"), 3, "L1T3 である");

  const temporalIds = new Set<number>();
  let keyFrames = 0;
  for (const frame of frames) {
    assert.ok(isRecord(frame));
    temporalIds.add(numberOf(frame, "temporalId"));
    if (boolOf(frame, "keyFrame")) {
      keyFrames += 1;
    }
    assert.ok(textOf(frame, "payloadHex").length > 0, "ペイロードが空でない");
  }
  assert.deepEqual([...temporalIds].sort(), [0, 1, 2], "時間層 0〜2 が現れる");
  assert.ok(keyFrames >= 1, "キーフレームを含む");
});

test("実 AV1 のワイヤ符号化が資産のバイト列と一致し、復号が往復する", () => {
  const asset = readAsset();
  const senderId = numberOf(asset, "senderId");
  const video = asset["video"];
  assert.ok(isRecord(video));
  const frames = video["frames"];
  assert.ok(Array.isArray(frames));
  const framerate = numberOf(video, "framerate");

  let checked = 0;
  for (const frame of frames) {
    assert.ok(isRecord(frame));
    const sequenceNumber = numberOf(frame, "sequenceNumber");
    const payload = hexToBytes(textOf(frame, "payloadHex"));
    const message: MediaMessage = {
      channel: CHANNEL_VIDEO,
      senderId,
      units: [
        {
          sequenceNumber,
          captureTimestampUs: BigInt(Math.trunc(((sequenceNumber - 1) * 1_000_000) / framerate)),
          flags: numberOf(frame, "expectedFlags"),
          spatialId: numberOf(frame, "spatialId"),
          temporalId: numberOf(frame, "temporalId"),
          payload,
        },
      ],
    };
    const encoded = encodeMediaMessage(message);
    assert.equal(encoded.ok, true, "符号化できる");
    if (!encoded.ok) {
      return;
    }
    assert.equal(bytesToHex(encoded.value), textOf(frame, "expectedMessageHex"), `映像 ${sequenceNumber} のバイト列`);

    const decoded = decodeMediaMessage(encoded.value);
    assert.equal(decoded.ok, true, "復号できる");
    if (!decoded.ok) {
      return;
    }
    const unit = decoded.value.units[0];
    assert.ok(unit !== undefined);
    assert.equal(bytesToHex(unit?.payload ?? new Uint8Array()), textOf(frame, "payloadHex"), "ペイロードが往復する");
    checked += 1;
  }
  assert.ok(checked >= 30, `30 枚以上を照合した（実際 ${checked}）`);
});

test("実 AV1 の破棄可否と破棄優先順位が資産と一致する", () => {
  const asset = readAsset();
  const video = asset["video"];
  assert.ok(isRecord(video));
  const frames = video["frames"];
  assert.ok(Array.isArray(frames));
  const temporalLayers = numberOf(video, "temporalLayers");

  let discardableCount = 0;
  for (const frame of frames) {
    assert.ok(isRecord(frame));
    const flags = numberOf(frame, "expectedFlags");
    const discardable = computeDiscardable(
      CHANNEL_VIDEO,
      boolOf(frame, "keyFrame"),
      numberOf(frame, "temporalId"),
      temporalLayers,
    );
    assert.equal(discardable, boolOf(frame, "expectedDiscardable"), "破棄可否が一致する");
    if (discardable) {
      discardableCount += 1;
    }
    const priority = dropPriority(CHANNEL_VIDEO, flags);
    const expected = frame["expectedDropPriority"];
    if (expected === null) {
      assert.equal(priority, null, "破棄禁止が一致する");
    } else {
      assert.equal(priority, expected, "優先順位が一致する");
    }
  }
  // 最上位の時間層は破棄可能である。1 枚も無ければ判断を検証していない。
  assert.ok(discardableCount > 0, `破棄可能なフレームがある（実際 ${discardableCount} 枚）`);
});

test("実 Opus の束ねが資産のバイト列と一致し、音声は破棄禁止である", () => {
  const asset = readAsset();
  const senderId = numberOf(asset, "senderId");
  const audio = asset["audio"];
  assert.ok(isRecord(audio));
  const bundles = audio["bundles"];
  assert.ok(Array.isArray(bundles));
  assert.ok(bundles.length >= 20, `音声束が 20 個以上ある（実際 ${bundles.length}）`);
  const frameMs = numberOf(audio, "frameMs");
  const unitsPerMessage = numberOf(audio, "unitsPerMessage");

  let checked = 0;
  for (let index = 0; index < bundles.length; index += 1) {
    const bundle = bundles[index];
    assert.ok(isRecord(bundle));
    const payloads = bundle["payloadsHex"];
    assert.ok(Array.isArray(payloads));
    assert.equal(payloads.length, unitsPerMessage, "束ねる数が規範どおりである");
    const first = numberOf(bundle, "firstSequenceNumber");
    const flags = numberOf(bundle, "expectedFlags");

    const units = payloads.map((payloadHex, offset) => {
      assert.equal(typeof payloadHex, "string");
      const hex = typeof payloadHex === "string" ? payloadHex : "";
      return {
        sequenceNumber: first + offset,
        captureTimestampUs: BigInt((index * unitsPerMessage + offset) * frameMs * 1000),
        flags,
        spatialId: 0,
        temporalId: 0,
        payload: hexToBytes(hex),
      };
    });
    const encoded = encodeMediaMessage({ channel: CHANNEL_AUDIO, senderId, units });
    assert.equal(encoded.ok, true, "符号化できる");
    if (!encoded.ok) {
      return;
    }
    assert.equal(bytesToHex(encoded.value), textOf(bundle, "expectedMessageHex"), `音声束 ${index} のバイト列`);

    const decoded = decodeMediaMessage(encoded.value);
    assert.equal(decoded.ok, true, "復号できる");
    if (!decoded.ok) {
      return;
    }
    assert.equal(decoded.value.units.length, unitsPerMessage, "ユニット数が往復する");

    // 音声は決して破棄しない（規範）。
    assert.equal(dropPriority(CHANNEL_AUDIO, flags), null, "音声は破棄禁止である");
    assert.equal(computeDiscardable(CHANNEL_AUDIO, false, 0, 1), false, "音声は破棄可能にならない");
    checked += 1;
  }
  assert.ok(checked >= 20, `20 束以上を照合した（実際 ${checked}）`);
});

File: sdks/dart/test/real_media_test.dart
// 実音声映像の照合（Dart、段 A の実データ版）。
//
// 合成したベクタではなく、**実際に符号化された AV1 と Opus**（spec/vectors/real-media.json）に
// 対して、ワイヤ形式の符号化・復号が往復で一致し、破棄可否と破棄優先順位の判断が
// 凍結資産と一致することを確かめる。同じ資産を 6 言語すべてが照合する。
//
// 資産を実装に合わせて変更してはならない（ADR-0012）。
// 動的型（dynamic）を使わない。JSON は Object? で受け、is で絞る。

import 'dart:convert';
import 'dart:io';
import 'dart:typed_data';

import 'package:test/test.dart';
import 'package:wheso_client/src/wire.dart';

Map<String, Object?> readMap(Object? value, String where) {
  if (value is Map<String, Object?>) {
    return value;
  }
  throw StateError('$where: 連想配列ではない');
}

List<Object?> readList(Object? value, String where) {
  if (value is List<Object?>) {
    return value;
  }
  throw StateError('$where: 配列ではない');
}

int readInt(Map<String, Object?> map, String key, String where) {
  final value = map[key];
  if (value is int) {
    return value;
  }
  throw StateError('$where: $key が整数ではない');
}

String readString(Map<String, Object?> map, String key, String where) {
  final value = map[key];
  if (value is String) {
    return value;
  }
  throw StateError('$where: $key が文字列ではない');
}

bool readBool(Map<String, Object?> map, String key, String where) {
  final value = map[key];
  if (value is bool) {
    return value;
  }
  throw StateError('$where: $key が真偽値ではない');
}

Uint8List hexToBytes(String hex) {
  final bytes = Uint8List(hex.length ~/ 2);
  for (var index = 0; index + 1 < hex.length; index += 2) {
    bytes[index ~/ 2] = int.parse(hex.substring(index, index + 2), radix: 16);
  }
  return bytes;
}

String bytesToHex(Uint8List bytes) {
  final builder = StringBuffer();
  for (final byte in bytes) {
    builder.write(byte.toRadixString(16).padLeft(2, '0'));
  }
  return builder.toString();
}

Map<String, Object?> readAsset() {
  final file = File('../../spec/vectors/real-media.json');
  return readMap(jsonDecode(file.readAsStringSync()), 'real-media.json');
}

void main() {
  test('実 AV1 のワイヤ符号化が資産と一致し、復号が往復する', () {
    final asset = readAsset();
    final senderId = readInt(asset, 'senderId', '資産');
    final video = readMap(asset['video'], 'video');
    final framerate = readInt(video, 'framerate', 'video');
    final channel = readInt(video, 'channel', 'video');
    final frames = readList(video['frames'], 'frames');
    expect(frames.length, greaterThanOrEqualTo(30), reason: '映像が 30 枚以上ある');

    var checked = 0;
    for (final raw in frames) {
      final frame = readMap(raw, 'frame');
      final sequenceNumber = readInt(frame, 'sequenceNumber', 'frame');
      final payloadHex = readString(frame, 'payloadHex', 'frame');
      final built = MediaMessage(
        channel: channel,
        senderId: senderId,
        units: <Unit>[
          Unit(
            sequenceNumber: sequenceNumber,
            captureTimestampUs: (sequenceNumber - 1) * 1000000 ~/ framerate,
            flags: readInt(frame, 'expectedFlags', 'frame'),
            spatialId: readInt(frame, 'spatialId', 'frame'),
            temporalId: readInt(frame, 'temporalId', 'frame'),
            payload: hexToBytes(payloadHex),
          ),
        ],
      );
      final encoded = encodeMediaMessage(built);
      expect(encoded.isOk, isTrue, reason: '符号化できる');
      final bytes = encoded.value;
      if (bytes == null) {
        throw StateError('符号化の結果が無い');
      }
      expect(bytesToHex(bytes), equals(readString(frame, 'expectedMessageHex', 'frame')),
          reason: '映像 $sequenceNumber のバイト列が資産と一致する');

      final decoded = decodeMediaMessage(bytes);
      expect(decoded.isOk, isTrue, reason: '復号できる');
      final message = decoded.value;
      if (message == null) {
        throw StateError('復号の結果が無い');
      }
      expect(bytesToHex(message.units[0].payload), equals(payloadHex), reason: 'ペイロードが往復する');
      checked += 1;
    }
    expect(checked, greaterThanOrEqualTo(30));
  });

  test('実 AV1 の破棄可否と破棄優先順位が資産と一致する', () {
    final asset = readAsset();
    final video = readMap(asset['video'], 'video');
    final channel = readInt(video, 'channel', 'video');
    final temporalLayers = readInt(video, 'temporalLayers', 'video');
    final frames = readList(video['frames'], 'frames');

    var discardableCount = 0;
    for (final raw in frames) {
      final frame = readMap(raw, 'frame');
      final discardable = computeDiscardable(
        channel,
        readBool(frame, 'keyFrame', 'frame'),
        readInt(frame, 'temporalId', 'frame'),
        temporalLayers,
      );
      expect(discardable, equals(readBool(frame, 'expectedDiscardable', 'frame')),
          reason: '破棄可否が資産と一致する');
      if (discardable) {
        discardableCount += 1;
      }
      final priority = dropPriority(channel, readInt(frame, 'expectedFlags', 'frame'));
      expect(priority, equals(frame['expectedDropPriority']), reason: '優先順位が資産と一致する');
    }
    // 最上位の時間層は破棄可能である。1 枚も無ければ判断を検証していない。
    expect(discardableCount, greaterThan(0), reason: '破棄可能なフレームがある');
  });

  test('実 Opus の束ねが資産と一致し、音声は破棄禁止である', () {
    final asset = readAsset();
    final senderId = readInt(asset, 'senderId', '資産');
    final audio = readMap(asset['audio'], 'audio');
    final frameMs = readInt(audio, 'frameMs', 'audio');
    final unitsPerMessage = readInt(audio, 'unitsPerMessage', 'audio');
    final channel = readInt(audio, 'channel', 'audio');
    final bundles = readList(audio['bundles'], 'bundles');
    expect(bundles.length, greaterThanOrEqualTo(20), reason: '音声束が 20 個以上ある');

    var checked = 0;
    for (var index = 0; index < bundles.length; index += 1) {
      final bundle = readMap(bundles[index], 'bundle');
      final payloads = readList(bundle['payloadsHex'], 'payloadsHex');
      expect(payloads.length, equals(unitsPerMessage), reason: '束ねる数が規範どおりである');
      final first = readInt(bundle, 'firstSequenceNumber', 'bundle');
      final flags = readInt(bundle, 'expectedFlags', 'bundle');

      final units = <Unit>[];
      for (var offset = 0; offset < payloads.length; offset += 1) {
        final payload = payloads[offset];
        if (payload is! String) {
          throw StateError('payloadsHex の要素が文字列ではない');
        }
        units.add(Unit(
          sequenceNumber: first + offset,
          captureTimestampUs: (index * unitsPerMessage + offset) * frameMs * 1000,
          flags: flags,
          spatialId: 0,
          temporalId: 0,
          payload: hexToBytes(payload),
        ));
      }
      final encoded = encodeMediaMessage(
        MediaMessage(channel: channel, senderId: senderId, units: units),
      );
      expect(encoded.isOk, isTrue, reason: '符号化できる');
      final bytes = encoded.value;
      if (bytes == null) {
        throw StateError('符号化の結果が無い');
      }
      expect(bytesToHex(bytes), equals(readString(bundle, 'expectedMessageHex', 'bundle')),
          reason: '音声束 $index のバイト列が資産と一致する');

      final decoded = decodeMediaMessage(bytes);
      expect(decoded.isOk, isTrue, reason: '復号できる');
      expect(decoded.value?.units.length, equals(unitsPerMessage), reason: 'ユニット数が往復する');

      // 音声は決して破棄しない（規範）。
      expect(dropPriority(channel, flags), isNull, reason: '音声は破棄禁止である');
      expect(computeDiscardable(channel, false, 0, 1), isFalse, reason: '音声は破棄可能にならない');
      checked += 1;
    }
    expect(checked, greaterThanOrEqualTo(20));
  });
}
